src/charm.py

```diff
@@ -336,12 +336,12 @@ def _on_get_certificate_action(self, event: ActionEvent) -> None:
         if not tls_certificates_relation:
             event.fail("Certificates relation not created.")
             return
-        if tls_certificates_relation.data[self.unit]["certificate"]:
+        if tls_certificates_relation.data[self.app]["certificate"]:
```

Comment:
Nitpick, this will be easier to to read if you extract `tls_certificates_relation.data[self.app]` to a variable. Same for the rest of the methods